src/HelicityAngles.cxx

```diff
@@ -9,109 +9,70 @@
 #include "ParticleCombination.h"
 #include "StatusManager.h"
 
-namespace yap {
+#include "logging.h"
 
-//-------------------------
-HelicityAngles::HelicityAngles(Model& m) :
-    StaticDataAccessor(m, equal_up_and_down),
-    Phi_(RealCachedValue::create(*this)),
-    Theta_(RealCachedValue::create(*this))
-{
-    registerWithModel();
-}
+#include <assert.h>
 
-//-------------------------
-void HelicityAngles::addToStaticDataAccessors()
-{
-    // look for Model's FourMomenta_
-    auto it_fm = std::find(staticDataAccessors().begin(), staticDataAccessors().end(), model()->fourMomenta().get());
-    if (it_fm == staticDataAccessors().end())
-        throw exceptions::Exception("HelicityAngles cannot be registered with the model before FourMomenta", "HelicityAngles::registerWithModel");
-    // add this to just after FourMomenta_
-    const_cast<StaticDataAccessorVector&>(staticDataAccessors()).insert(it_fm + 1, this);
-}
+namespace yap {
 
-//-------------------------
-double HelicityAngles::phi(const DataPoint& d, const std::shared_ptr<const ParticleCombination>& pc) const
-{
-    return Phi_->value(d, symmetrizationIndex(pc));
-}
 
 //-------------------------
-double HelicityAngles::theta(const DataPoint& d, const std::shared_ptr<const ParticleCombination>& pc) const
+const std::array<double, 2>& HelicityAngles::helicityAngles(const DataPoint& d, const StatusManager& sm, const std::shared_ptr<const ParticleCombination>& pc) const
 {
-    return Theta_->value(d, symmetrizationIndex(pc));
-}
+    // check if DataPoint is currently in cache
+    if (cachedForDataPoint_[&sm] == &d) {
+        // find entry
+        for (const auto& kv : cachedAngles_[&sm])
+            if (equal_up_and_down(kv.first, pc))
+                return kv.second;
+    }
+    else {
+        // reset and clear
+        cachedForDataPoint_[&sm] = nullptr;
+        cachedAngles_[&sm].clear();
+    }
 
-//-------------------------
-void HelicityAngles::calculate(DataPoint& d, StatusManager& sm) const
-{
-    // set angles uncalculated
-    sm.set(*this, CalculationStatus::uncalculated);
-
-    // call on ISP PC's
-    // \todo allow for designating the boost that takes from the data frame to the lab frame (possibly event dependent)
-    for (auto& kv : symmetrizationIndices())
-        if (not kv.first->parent())
-            calculateAngles(d, kv.first, model()->coordinateSystem(), unitMatrix<double, 4>(), sm);
+    // if not found, calculate
+    calculateAngles(d, sm, origin(*pc).shared_from_this(), Model_->coordinateSystem(), unitMatrix<double, 4>());
+    cachedForDataPoint_[&sm] = &d;
+
+    return helicityAngles(d, sm, pc);
```

Comment:
If something were to go wrong, wouldn't this infinitely loop? It might be better to just now retrieve the calculated the value or throw if it somehow wasn't calculated in the above lines.